letsencrypt/client.py

```diff
@@ -30,48 +31,107 @@
+    :returns: Newly registered and saved account, as well as protocol
+        API handle (should be used in `Client` initialization).
+    :rtype: `tuple` of `.Account` and `acme.client.Client`
 
     """
+    # Log non-standard actions, potentially wrong API calls
+    if account_storage.find_all():
+        logger.info("There are already existing accounts for %s", config.server)
+    if config.email is None:
+        logger.warn("Registering without email!")
+
+    # Each new registration shall use a fresh new key
+    key = jose.JWKRSA(key=jose.ComparableRSAKey(
+        rsa.generate_private_key(
+            public_exponent=65537,
+            key_size=config.rsa_key_size,
+            backend=default_backend())))
+    acme = _acme_from_config_key(config, key)
+    # TODO: add phone?
+    regr = acme.register(messages.Registration.from_data(email=config.email))
+
+    if regr.terms_of_service is not None:
+        if tos_cb is not None and not tos_cb(regr):
+            raise errors.Error(
+                "Registration cannot proceed without accepting "
+                "Terms of Service.")
+        regr = acme.agree_to_tos(regr)
+
+    acc = account.Account(regr, key)
+    account.report_new_account(acc, config)
+    account_storage.save(acc)
+    return acc, acme
+
+
+class Client(object):
+    """ACME protocol client.
 
-    def __init__(self, config, account_, dv_auth, installer):
+    :ivar .IConfig config: Client configuration.
+    :ivar .Account account: Account registered with `register`.
+    :ivar .AuthHandler auth_handler: Authorizations handler that will
+        dispatch DV and Continuity challenges to appropriate
+        authenticators (providing `.IAuthenticator` interface).
+    :ivar .IInstaller installer: Installer.
+    :ivar acme.client.Client acme: Optional ACME client API handle.
+       You might already have one from `register`.
+
+    """
+
+    def __init__(self, config, account_, dv_auth, installer, acme=None):
         """Initialize a client.
 
-        :param dv_auth: IAuthenticator that can solve the
-            :const:`letsencrypt.constants.DV_CHALLENGES`.
-            The :meth:`~letsencrypt.interfaces.IAuthenticator.prepare`
-            must have already been run.
-        :type dv_auth: :class:`letsencrypt.interfaces.IAuthenticator`
+        :param .IAuthenticator dv_auth: Prepared (`.IAuthenticator.prepare`)
+            authenticator that can solve the `.constants.DV_CHALLENGES`.
 
         """
+        self.config = config
         self.account = account_
-
         self.installer = installer
 
-        # TODO: Allow for other alg types besides RS256
-        self.network = network.Network(
-            config.server, jwk.JWKRSA.load(self.account.key.pem),
-            verify_ssl=(not config.no_verify_ssl))
-
-        self.config = config
+        # Initialize ACME if account is provided
+        if acme is None and self.account is not None:
```

Comment:
You need an Account Key to create acme. In fact we should have `assert account is not None` here, but this requires a whole `client.py` rewrite that I'll be working on next :)